Backtrace Debug tweaks

dtolnay opened the pull request:
- Change Debug representation of disabled and unsupported backtraces to use \<placeholder\> style, same as what we do for debug printing locked mutexes and mutably borrowed refcells;

    ```diff
    - Error { msg: "...", backtrace: disabled backtrace }
    + Error { msg: "...", backtrace: <disabled> }
    ```

- Remove quotes around unresolved symbol names;

    ```diff
    - Backtrace [{ fn: "<unknown>" }]
    + Backtrace [{ fn: <unknown> }]
    ```

- Add quotes around file paths;

    ```diff
    - Backtrace [{ fn: "krate::main", file: /path/to/main.rs, line: 10 }]
    + Backtrace [{ fn: "krate::main", file: "/path/to/main.rs", line: 10 }]
    ```

- Add test.

dtolnay committed:
Remove quotes around unknown fn placeholder in backtrace

dtolnay committed:
Add quotes around filename in Backtrace debug

dtolnay committed:
Change disabled and unsupported backtraces to print using placeholder style

dtolnay committed:
Make it possible to instantiate hardcoded Backtrace from test

dtolnay committed:
Add test of Debug representation of Backtrace

rust-highfive commented:
r? @cramertj

(rust_highfive has picked a reviewer for you, use r? to override)

cramertj commented:
This looks like a good improvement to me, thanks!

@bors r+

bors commented:
:pushpin: Commit 3e4992bd0bdce396b31dbc74adb6bb189e60dc72 has been approved by `cramertj`

<!-- @bors r=cramertj 3e4992bd0bdce396b31dbc74adb6bb189e60dc72 -->
<!-- homu: {"type":"Approved","sha":"3e4992bd0bdce396b31dbc74adb6bb189e60dc72","approver":"cramertj"} -->

bors commented:
:hourglass: Testing commit 3e4992bd0bdce396b31dbc74adb6bb189e60dc72 with merge 805bc0a427640f64b9690d2c1c697338e907c6f2...
<!-- homu: {"type":"BuildStarted","head_sha":"3e4992bd0bdce396b31dbc74adb6bb189e60dc72","merge_sha":"805bc0a427640f64b9690d2c1c697338e907c6f2"} -->

rust-highfive commented:
The job `armhf-gnu` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=20942&view=logs&jobId=bedce312-f790-5a16-cf2f-66c91a66d998) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20942/logs/1609), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/20942/logs/1609)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-02-14T07:41:31.2466964Z failures:
2020-02-14T07:41:31.2467188Z 
2020-02-14T07:41:31.2468151Z ---- backtrace::test_debug stdout ----
2020-02-14T07:41:31.2468852Z thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
2020-02-14T07:41:31.2469315Z   left: `"Backtrace [\n    { fn: \"__rust_maybe_catch_panic\" },\n    { fn: \"std::rt::lang_start_internal\", file: \"./rust/rt.rs\", line: 300 },\n    { fn: \"std::rt::lang_start\", file: \"./rust/rt.rs\", line: 400 },\n]"`,
2020-02-14T07:41:31.2470232Z  right: `"Backtrace [\n    { fn: \"__rust_maybe_catch_panic\" },\n    { fn: \"std::rt::lang_start_internal\", file: \"/rust/rt.rs\", line: 300 },\n    { fn: \"std::rt::lang_start\", file: \"/rust/rt.rs\", line: 400 },\n]"`', src/libstd/backtrace.rs:486:5
2020-02-14T07:41:31.2470946Z 
2020-02-14T07:41:31.2471179Z 
2020-02-14T07:41:31.2471807Z failures:
2020-02-14T07:41:31.2472061Z     backtrace::test_debug
---
2020-02-14T07:41:31.2596857Z 
2020-02-14T07:41:31.2596898Z 
2020-02-14T07:41:31.2597554Z failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --target arm-unknown-linux-gnueabihf
2020-02-14T07:41:31.2597823Z Build completed unsuccessfully in 2:45:22
2020-02-14T07:41:31.2598461Z error: test failed, to rerun pass '-p std --lib'
2020-02-14T07:41:31.2659241Z   local time: Fri Feb 14 07:41:31 UTC 2020
2020-02-14T07:41:31.5540902Z   network time: Fri, 14 Feb 2020 07:41:31 GMT
2020-02-14T07:41:31.5545990Z == end clock drift check ==
2020-02-14T07:41:32.0895147Z 
2020-02-14T07:41:32.0895147Z 
2020-02-14T07:41:32.0990657Z ##[error]Bash exited with code '1'.
2020-02-14T07:41:32.1038944Z ##[section]Starting: Checkout rust-lang/rust@auto to s
2020-02-14T07:41:32.1041106Z ==============================================================================
2020-02-14T07:41:32.1041222Z Task         : Get sources
2020-02-14T07:41:32.1041306Z Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=20942)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=20942","builder_name":"checks-azure"} -->

joelpalmer commented:
Triaged

dtolnay committed:
Write backtrace fmt test using relative paths

For some reason the absolute paths were formatted differently on the
armhf-gnu target.

    thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
      left: `"Backtrace [\n    { fn: \"__rust_maybe_catch_panic\" },\n    { fn: \"std::rt::lang_start_internal\", file: \"./rust/rt.rs\", line: 300 },\n    { fn: \"std::rt::lang_start\", file: \"./rust/rt.rs\", line: 400 },\n]"`,
     right: `"Backtrace [\n    { fn: \"__rust_maybe_catch_panic\" },\n    { fn: \"std::rt::lang_start_internal\", file: \"/rust/rt.rs\", line: 300 },\n    { fn: \"std::rt::lang_start\", file: \"/rust/rt.rs\", line: 400 },\n]"`', src/libstd/backtrace.rs:486:5

dtolnay commented:
- Rebase
- Fix armhf-gnu build

@bors r=cramertj

bors commented:
:pushpin: Commit a2364dc85fb505487c272988b8c5073a3f6fef2a has been approved by `cramertj`

<!-- @bors r=cramertj a2364dc85fb505487c272988b8c5073a3f6fef2a -->
<!-- homu: {"type":"Approved","sha":"a2364dc85fb505487c272988b8c5073a3f6fef2a","approver":"cramertj"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 1000, this pull request will be tested once the tree is reopened
